### DISCUSSION TITLE
How to pass object data to custom scripts via API call?

### DISCUSSION THREAD

**Discussion by zshorvat** [netbox-community/netbox#8344] [Q&A] [answered]
Python 3.9.2 | Django 3.2.9 | NetBox 3.0.10

I'm trying to call my custom scripts via the REST API, and in the [documentation](https://netbox.readthedocs.io/en/stable/customization/custom-scripts/#via-the-api) you explain how to pass simple form variables, but I'm rather stuck when it comes to objects.

I'm trying to pass a device object to my custom script in python via the API, tried to squeeze the whole device dict into the data JSON, also tried passing only the device ID but in both cases the custom script fails to pick up attributes of the expected object.

It might also be possible that I have something dodgy in my python code, so pasted the relevant part below:

```
 nbheaders = {
     "Authorization" : f"Token {nbapikey}",
     "Content-Type" : "application/json"
 }

 configcontext_data={}
 configcontext_data["data"]={}
 configcontext_data["data"]["hosts"] = nb_device_id
 configcontext_data["data"]["loglevel"] =  "NONE"
 configcontext_data["commit"] = True
 
 configcontext_update = requests.post("https://<netboxserver>/api/extras/scripts/prod_createConfigContext.createConfigContext/", headers=nbheaders, data=json.dumps(configcontext_data), verify=False)
```
Probably for others' benefit as well, it might be useful to call out on the doc page if there are any tips or tricks as most people I think will use objects in their custom scripts.

Thanks,
Zsolt

**Comment by candlerb** [comment]
Can you show how `hosts` is defined in your script? Is it an ObjectVar, a MultiObjectVar, or something else?

**Reply by zshorvat** [reply]
Not sure why my prev reply didn't come in, but anyway - many thanks for looking into this topic!

This is how my custom script is defined:

```python
class createConfigContext(Script):
    class Meta:
        name = "Create device config context"
        description = "Create configuration used for ACI configuration"
        field_order = ['hostname']
        commit_default = False

    hosts = ObjectVar(
        description="Host",
        model=Device,
        required=True,
        query_params={
            'status': 'active'
        }
    )

    loglevel_list = (
        ('NONE', 'No logging'),
        ('INFO', 'Informational'),
        ('DEBUG', 'Debug'),
    )
    loglevel = ChoiceVar(
        choices=loglevel_list, 
        default='INFO', 
        description="Loglevel required"
    )

```

**Comment by zshorvat** [comment]


**Reply by candlerb** [reply]
It looks like a missing feature to be able to pass objects via the API.  Complete test:

```
from extras.scripts import Script, ObjectVar
from dcim.models import Device
from django.utils.html import escape

class CheckObjectVar(Script):
    class Meta:
        name = "Check ObjectVar"
        commit_default = False

    hosts = ObjectVar(
        description="Host",
        model=Device,
        required=True,
        query_params={
            'status': 'active'
        }
    )

    def run(self, data, commit):
        return "device: %s" % escape(repr(data["hosts"]))
```

Whatever I pass as `{"data": {"hosts":xxx}, "commit": false}` just gets reflected as a literal value - a string, a number, whatever.

**Reply by zshorvat** [reply]
Thanks, that helped a lot!

I worked around the issue by adding a new key "host_id" to the data dict - if it is present, I look up my host instead of trying to read it from data directly.

Thanks again,
Zsolt

**Comment by zshorvat** [comment, marked as answer]
For the github discussion's sake: the feature is missing, but it is possible to work it around by passing the ID under another key in "data" when calling the API. If that key is present, lookup should be done before trying to do anything with the host object.

**Reply by LuPo** [reply]
I can not figure out how exactly you passed the object id to the data. I am also trying to use MultiObjectVar instead of ObjectVar.

**Reply by mathieumd** [reply]
I don't know how @zshorvat did it, but here is how I did to get a list of Devices from a call to API via `curl`, providing a list of Devices IDs, a Rack ID or a Site ID (look at the 3 `devices_id`, `rack_id` and `site_id`):

```py
    def run(self, data, commit):
        selected_devices = []
        if 'devices_id' in data:
            for d in data['devices_id']:
                selected_devices.append(Device.objects.get(id=d))
        elif 'rack_id' in data:
            selected_devices = Device.objects.filter(rack__id=data['rack_id'])
        elif 'site_id' in data:
            selected_devices = Device.objects.filter(site__id=data['site_id'])
        elif 'devices' in data and data['devices']:
            selected_devices = data['devices']
        elif 'rack' in data and data['rack']:
            selected_devices = Device.objects.filter(rack=data['rack'])
        elif 'site' in data and data['site']:
            selected_devices = Device.objects.filter(site=data['site'])

        if len(selected_devices) == 0:
            self.log_warning("Aucun device trouvé !")
            return

        #...
```

Curl call:

```sh
curl ... --json '{"data": {"rack_id": 4}, "commit": false}'
```
